### ISSUE TITLE
need misc updates for dev testing documentation

### ISSUE THREAD

**Issue by mattvonrocketstein** [ipython/ipython#7127]
This issue refers to the documentation here: https://github.com/ipython/ipython/wiki/Dev%3A-Testing

Hopefully I'm looking at the right stuff.  One example mentioned is 

```
$ iptest -v IPython.utils
..........................SS..SSS............................S.S...
.........................................................
----------------------------------------------------------------------
 Ran 125 tests in 0.119s
```

This doesn't work for me, because it should be `iptest IPythin.utils -- -v`.  I see

```
> iptest -v IPython.utils
usage: iptest [-h] [--all] [--slimerjs] [--url URL] [-j [FAST]] [--xunit] [--coverage [COVERAGE]]
              [--subproc-streams SUBPROC_STREAMS]
              [testgroups [testgroups ...]]
iptest: error: unrecognized arguments: -v
```

Another broken example is `python tools/test_pr.py -p 1234`, which broke for me because I didn't have the keyring module installed.  Requirements are mentioned in the documentation nearby, but shouldn't this be installed as part of "ipython[all]" requirements?  Once I installed keyring, the example still didn't work.  Does this run on the official travis, and is only intended for core committers with extra privs, or can I expect to do this on my local env?  I see:

```
 > python ipython/tools/test_pr.py -p 7125
no cache
Please enter your github username and password. These are not stored, only used to get an oAuth token. You can revoke this at any time on Github.
Username: mattvonrocketstein
Password: 
Traceback (most recent call last):
  File "ipython/tools/test_pr.py", line 289, in <module>
    test_pr(args.number, post_results=args.publish, extra_args=extra_args)
  File "ipython/tools/test_pr.py", line 259, in test_pr
    gh_api.get_auth_token()
  File "/home/vagrant/code/ipython/tools/gh_api.py", line 67, in get_auth_token
    response.raise_for_status()
  File "/home/vagrant/code/ipython/ipy_venv/local/lib/python2.7/site-packages/requests/models.py", line 825, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 422 Client Error: Unprocessable Entity
```

Finally and probably most importantly, I think a section on testing extensions is missing.  Extension writers can unittest their own stuff however they wish, but obviously the more integration-style tests used in ipython core could be a very useful for complex extensions.  Are there any existing thoughts about best practice here or better yet examples of third-party code doing this?  Is iptest pretty agnostic about whether it's invoked on IPython.\* or MyExtension.*? etc

Thanks

**Comment by ellisonbg** [comment]
Any progress on this? Is there anything we need to do?

**Comment by mattvonrocketstein** [comment]
Needs info from me you mean?  I'd say nothing has been resolved.  Trying to make things more concrete:

1. Does test_pr.py get to stay or does it go?  
2. Should there or should there not be documentation on strategies for writing extension tests?
3. If there should be documentation for it, I'm trying to write it but I need some hints first.  

Unit testing is straightforward for third party apps since they can do whatever, but, how best to write and run IPython.core style integration tests outside of IPython development?

**Comment by ellisonbg** [comment]
Sorry, I just went through a couple hundred issues and ended up adding needs-info to issues that I wasn't quite sure what the state was. Thanks, your note helps to clarify what needs to be done though.

**Comment by ellisonbg** [comment]
I should note I am fine getting rid of test_pr. @fperez do you still use it?

**Comment by takluyver** [comment]
test_pr is theoretically still useful for things we don't test on Travis, like IPython.parallel, or if Travis is down/achingly slow. We don't often use it, however, and presumably there'll be even less need for it after the repo split.

**Comment by fperez** [comment]
Now that I have `git mpr` *and* most cases are taken care of by Travis, the need for `test_pr` is significantly lower. I'd be OK if it goes, if that helps simplify things.

**Comment by ivanov** [comment]
Hey there. I'm going through old issues and it seems to me that it makes sense to close this one. There were multiple threads here, some of which have been addressed: wiki was edited, and a decision was made that removing `test_pr.py` is fine (I've just posted #10041 to do just that). The last bit which wasn't addressed - which is to document strategies for writing extensions, is clouded by the other stuff. Documentation updates are always welcome, no point in having an issue open to state that (and in a roundabout manner.)
